src/manager_service.py

```diff
@@ -0,0 +1,146 @@
+    except SubprocessError as err:
+        raise RunnerManagerApplicationInstallError(_INSTALL_ERROR_MESSAGE) from err
+
+    logger.info("Uninstalling previous version of packages")
+    try:
+        execute_command(["python3", "-m", "pip", "uninstall", GITHUB_RUNNER_MANAGER_PACKAGE])
+        execute_command(["python3", "-m", "pip", "uninstall", JOB_MANAGER_PACKAGE])
+    except SubprocessError:
+        logger.info(
+            "Unable to uninstall existing packages, likely due to previous version not installed"
+        )
+
+    try:
+        # Use `--prefix` to install the package in a location (/usr) all user can use and
+        # `--ignore-installed` to force all dependencies be to installed under /usr.
+        execute_command(
+            [
+                "python3",
+                "-m",
+                "pip",
+                "install",
+                "--prefix",
+                "/usr",
+                "--ignore-installed",
+                GITHUB_RUNNER_MANAGER_PACKAGE_PATH,
+                JOB_MANAGER_PACKAGE_PATH,
+            ]
+        )
+    except SubprocessError as err:
+        raise RunnerManagerApplicationInstallError(_INSTALL_ERROR_MESSAGE) from err
+
+
+def _enable_service() -> None:
+    """Enable the github runner manager service.
+
+    Raises:
+        RunnerManagerApplicationStartupError: Unable to startup the service.
+    """
+    try:
+        systemd.service_enable(GITHUB_RUNNER_MANAGER_SERVICE_NAME)
+        if not systemd.service_running(GITHUB_RUNNER_MANAGER_SERVICE_NAME):
+            systemd.service_start(GITHUB_RUNNER_MANAGER_SERVICE_NAME)
+    except SystemdError as err:
+        raise RunnerManagerApplicationStartupError(_SERVICE_SETUP_ERROR_MESSAGE) from err
+
+
+def _setup_config_file(state: CharmState, app_name: str, unit_name: str) -> Path:
+    """Write the configuration to file.
+
+    Args:
+        state: The charm state.
+        app_name: The Juju application name.
+        unit_name: The Juju unit.
+    """
+    config = create_application_configuration(state, app_name, unit_name)
+    # Directly converting to `dict` will have the value be Python objects rather than string
+    # representations. The values needs to be string representations to be converted to YAML file.
+    # No easy way to directly convert to YAML file, so json module is used.
+    config_dict = json.loads(config.json())
+    path = Path(f"~{constants.RUNNER_MANAGER_USER}").expanduser() / "config.yaml"
+    with open(path, "w+", encoding="utf-8") as file:
+        yaml_safe_dump(config_dict, file)
+    return path
+
+
+def _setup_service_file(config_file: Path) -> None:
+    """Configure the systemd service.
+
+    Args:
+        config_file: The configuration file for the service.
+    """
+    jinja = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+
+    service_file_content = jinja.get_template("github-runner-manager.service.j2").render(
```

Comment:
Since the template is trivial (no complex logic) and is pretty short, it's probably more readable to use a multiline f-string instead